sync-core/src/main/java/com/cloudant/sync/datastore/AttachmentManager.java

```diff
@@ -348,47 +382,172 @@ protected void purgeAttachments(SQLDatabase db) {
+            c = db.rawQuery(SQL_ATTACHMENTS_SELECT_KEYS_FILENAMES, null);
+            while (c.moveToNext()) {
+                String keyForFile = c.getString(0);
+
+                if (!currentKeys.contains(keyForFile)) {
+                    File f = new File(attachments, c.getString(1));
+                    try {
+                        boolean deleted = f.delete();
+                        if (deleted) {
+                            db.delete(ATTACHMENTS_KEY_FILENAME, "key = ?", new String[]{keyForFile});
+                        } else {
+                            logger.warning("Could not delete file from BLOB store: " +
+                                    f.getAbsolutePath());
                         }
+                    } catch (SecurityException e) {
+                        String msg = String.format("SecurityException deleting %s from blob store",
+                                f.getAbsolutePath());
+                        logger.log(Level.WARNING, msg, e);
                     }
                 }
-        }
+            }
+            c.close();
+
         } catch (SQLException e) {
-            logger.log(Level.SEVERE, "Problem in purgeAttachments, executing SQL to fetch all attachment keys ", e);
-        }finally {
-           DatabaseUtils.closeCursorQuietly(c);
+            logger.log(Level.SEVERE,
+                    "SQL exception in purgeAttachments when removing redundant attachments", e);
+        } finally {
+            DatabaseUtils.closeCursorQuietly(c);
         }
     }
 
-    private String keyToString(byte[] key) {
+    private static String keyToString(byte[] key) {
         return new String(new Hex().encode(key));
     }
 
-    private File fileFromKey(byte[] key) {
-        return new File(attachmentsDir, keyToString(key));
+    /**
+     * Lookup or create a on disk File representation of blob, in {@code db} using {@code key}.
+     *
+     * Existing attachments will have an entry in db, so the method just looks this up
+     * and returns the File object for the path.
+     *
+     * For new attachments, the {@code key} doesn't already have an associated filename,
+     * one is generated and inserted into the database before returning a File object
+     * for blob associated with {@code key}.
+     *
+     * @param db database to use.
+     * @param key key to lookup filename for.
+     * @param allowCreateName if the is no existing mapping, whether one should be created
+     *                        and returned. If false, and no existing mapping, AttachmentException
+     *                        is thrown.
+     * @return File object for blob associated with {@code key}.
+     * @throws AttachmentException if a mapping doesn't exist and {@code allowCreateName} is
+     *          false or if the name generation process fails.
+     */
+    File fileFromKey(SQLDatabase db, byte[] key, boolean allowCreateName) throws AttachmentException {
```

Comment:
I found the logic more easy to reason about if there was just one call -- allows to enforce a transaction around the "is there a key? no? generate one" logic.